### PR TITLE
feat: Add `Repo.flush` method

### PR DESCRIPTION
**Pull request by mykola-vrmchk**
## Motivation 
Currently, there is no way to know if auto merge `Repo` finished writing all updates to disk. In our use-case, we want to know when it is safe to shut down the process (e.g. CLI) to preserve all data on disk.

related https://github.com/automerge/automerge-repo/issues/264
related [discord discussion](https://discord.com/channels/1200006940210757672/1214251439438827520)

### CODE REVIEW

File: packages/automerge-repo/src/Repo.ts [hunk 1]
```diff
@@ -505,6 +505,31 @@ export class Repo extends EventEmitter<RepoEvents> {
...
+  /**
+   * Waits for Repo to finish write changes to disk.
+   * @deprecated
+   * @param documents - if provided, only waits for the specified documents
+   * @param timeout - if provided, the maximum time to wait in milliseconds
+   * @returns Promise<void>
+   */
+  async flush(documents?: DocumentId[], timeout?: number): Promise<void> {
+    if (!this.storageSubsystem) {
+      throw new Error("flush called without storage subsystem")
```
Comment (bijela-gora) [inline packages/automerge-repo/src/Repo.ts:518]: I checked other places in code, and it looks like it is OK to have `storageSubsystem` defined. The code should return resolved promise here.

File: packages/automerge-repo/src/Repo.ts [hunk 2]
```diff
@@ -505,6 +505,31 @@ export class Repo extends EventEmitter<RepoEvents> {
...
+      : Object.values(this.#handleCache)
+    return Promise.all(
+      handles.map(async handle => {
+        const doc = handle.docSync()
+        if (!doc) {
+          return
+        }
+        return this.storageSubsystem!.flush(handle.documentId, doc, timeout)
+      })
+    ).then(() => {})
```
Comment (bijela-gora) [inline packages/automerge-repo/src/Repo.ts:531]:
```suggestion
    ).then(() => {})
```
May I ask what this line is for?

File: packages/automerge-repo/src/Repo.ts [hunk 3]
```diff
@@ -505,6 +505,31 @@ export class Repo extends EventEmitter<RepoEvents> {
       return this.storageSubsystem.id()
     }
   }
+
+  /**
+   * Waits for Repo to finish write changes to disk.
+   * @deprecated
```
Comment (bijela-gora) [inline packages/automerge-repo/src/Repo.ts:511]: You marked a new code as deprecated?

File: packages/automerge-repo/src/Repo.ts [hunk 4]
```diff
@@ -505,6 +505,31 @@ export class Repo extends EventEmitter<RepoEvents> {
       return this.storageSubsystem.id()
     }
   }
+
+  /**
+   * Waits for Repo to finish write changes to disk.
```
Comment (bijela-gora) [inline packages/automerge-repo/src/Repo.ts:510]: Looks like a useful feature.

File: packages/automerge-repo/src/Repo.ts (hunk 3 above)
Comment (mykola-vrmchk) [reply packages/automerge-repo/src/Repo.ts:511]: I discussed it with @pvh. He said that his feeling is that `flush()` isn't quite the right API for external consumption but is internally useful. We agreed that it is ok to add it but I proposed to mark `flush` as deprecated to warn users that this API will probably change soon.

Comment (bijela-gora) [reply packages/automerge-repo/src/Repo.ts]: Nice! Thank you for the replay.

**Review by bijela-gora — COMMENTED** [review]
lgtm

**Comment by mykola-vrmchk** [comment]
> lgtm

@bijela-gora Could you please merge it, it is blocking us

**Comment by bijela-gora** [comment]
@mykola-vrmchk I have no such permissions. I reviewed your PR because it was interesting for me.

**Comment by bijela-gora** [comment]
@mykola-vrmchk I have no such permissions. I reviewed your PR because it was interesting for me.

**Pull request merged**